src/gc/gc.cpp

```diff
@@ -17284,6 +17284,94 @@ uint8_t* gc_heap::find_object (uint8_t* o, uint8_t* low)
 
 #define method_table(o) ((CObjectHeader*)(o))->GetMethodTable()
 
+inline
+BOOL gc_heap::is_string_and_about_to_be_promoted_to_gen2 (uint8_t* o)
+{
+    assert((object_gennum (o) == 1) == (
+        o != NULL &&
+        in_range_for_segment (o, ephemeral_heap_segment) &&
+        (o >= generation_allocation_start (generation_of (1))) &&
+        (o < generation_allocation_start (generation_of (0)))
+        ));
+    return (
+        o != NULL &&
+        settings.condemned_generation >= 1 &&
+        method_table (o) == *g_gc_pStringClass &&
+        in_range_for_segment (o, ephemeral_heap_segment) &&
+        (o >= generation_allocation_start (generation_of (1))) &&
+        (o < generation_allocation_start (generation_of (0))) &&
+        (((StringObject*)o)->GetStringLength() >= sizeof (uint8_t*) / 2)
+        );
+}
+
+inline
+BOOL gc_heap::try_relocate_duplicate_string (uint8_t** pold_address)
+{
+    uint8_t* old_address = *pold_address;
+    if (!old_address)
+    {
+        return false;
+    }
+    // it's sad we can't leverage mark bit to indicate that string is duplicate
```

Comment:
quick thought: the lowest two bits of the method table pointer are usable and we only use one of them. We could in theory use the other bit for this.